akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala

```diff
@@ -0,0 +1,122 @@
+import scala.collection.immutable
+import scala.util.Try
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object DnsClient {
+  sealed trait Question {
+    def id: Short
+  }
+  final case class SrvQuestion(id: Short, name: String) extends Question
+  final case class Question4(id: Short, name: String) extends Question
+  final case class Question6(id: Short, name: String) extends Question
+  final case class Answer(id: Short, rrs: immutable.Seq[DnsResourceRecord]) extends NoSerializationVerificationNeeded
+  final case class DropRequest(id: Short)
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] class DnsClient(ns: InetSocketAddress) extends Actor with ActorLogging with Stash {
+
+  import DnsClient._
+
+  import context.system
+
+  IO(Udp) ! Udp.Bind(self, new InetSocketAddress(InetAddress.getByAddress(Array.ofDim(4)), 0))
+
+  var requests: Map[Short, ActorRef] = Map.empty
+
+  def receive: Receive = {
+    case Udp.Bound(local) ⇒
+      log.debug(s"Bound to UDP address [{}]", local)
+      context.become(ready(sender()))
+      unstashAll()
+    case _: Question4 ⇒
+      stash()
+    case _: Question6 ⇒
+      stash()
+    case _: SrvQuestion ⇒
+      stash()
+  }
+
+  private def message(name: String, id: Short, recordType: DnsRecordType): Message = {
+    Message(id, MessageFlags(), immutable.Seq(DnsQuestion(name, recordType, DnsRecordClass.IN)))
+  }
+
+  def ready(socket: ActorRef): Receive = {
+    case DropRequest(id) ⇒
+      log.debug("Dropping request [{}]", id)
+      requests -= id
+    case Question4(id, name) ⇒
+      log.debug("Resolving [{}] (A)", name)
+      requests += (id -> sender())
+      val msg = message(name, id, DnsRecordType.A)
+      log.debug(s"Message [{}] to [{}]: [{}]", id, ns, msg)
+      socket ! Udp.Send(msg.write(), ns)
+
+    case Question6(id, name) ⇒
+      log.debug("Resolving [{}] (AAAA)", name)
+      requests += (id -> sender())
+      val msg = message(name, id, DnsRecordType.AAAA)
+      log.debug(s"Message to [{}]: [{}]", ns, msg)
+      socket ! Udp.Send(msg.write(), ns)
+
+    case SrvQuestion(id, name) ⇒
+      log.debug("Resolving [{}] (SRV)", name)
+      requests += (id -> sender())
+      val msg = message(name, id, DnsRecordType.SRV)
+      log.debug(s"Message to {}: msg", ns, msg)
+      socket ! Udp.Send(msg.write(), ns)
+
+    case Udp.CommandFailed(cmd) ⇒
+      log.debug("Command failed [{}]", cmd)
+      cmd match {
+        case send: Udp.Send ⇒
+          // best effort, don't throw
+          Try {
+            val msg = Message.parse(send.payload)
+            requests.get(msg.id).foreach { s ⇒
+              s ! Failure(new RuntimeException("Send failed to nameserver"))
+              requests -= msg.id
+            }
+          }
+        case _ ⇒
+          log.warning("Dns client failed to send {}", cmd)
+      }
+    case Udp.Received(data, remote) ⇒
+      log.debug(s"Received message from [{}]: [{}]", remote, data)
+      val msg = Message.parse(data)
```

Comment:
If response would be corrupted/bad and parsing fails and throws, does that stop the client?